### DISCUSSION TITLE
Got autoresearch running on an RTX 5070 12 GB

### DISCUSSION THREAD

**Discussion by fayerman-source** [karpathy/autoresearch#50] [Show and tell]
Posting this in case anyone else is trying to run `autoresearch` on more normal local NVIDIA hardware.

I tried it on an RTX 5070 12 GB box and the default path didn’t work cleanly for me as-is. The main problem was the attention path, plus some defaults that were clearly happier on bigger GPUs.

I made a fork here:
https://github.com/fayerman-source/autoresearch

What I changed:
- fallback to PyTorch SDPA on non-Hopper GPUs
- a few small presets:
  - `AR_PRESET=8gb`
  - `AR_PRESET=12gb`
  - `AR_PRESET=h100`
- clearer startup logging
- cleaned up the README so it’s easier to follow as an operator
- added a tiny helper for appending run summaries to `results.tsv`

What worked for me:
```bash
uv sync
uv run prepare.py
AR_PRESET=12gb uv run train.py
```

I also got a full 5-minute run to complete after those changes.

Not trying to turn it into a big compatibility layer. Just wanted a version that keeps the original shape but is less painful on a consumer card.

**Comment by guarinoindustries-ctrl** [comment]
Andrej didn't come up with this idea originally, this guy did: https://github.com/topherchris420/james_library

**Reply by guarinoindustries-ctrl** [reply]
Can he stop plagiarising? Alleged college grad, fails to cite his sources and work. Looks really bad on him. I'll keep spamming until he cites his work. I bet if it was your work, you'd love to be cited.

**Reply by guarinoindustries-ctrl** [reply]
Do you do any research whatsoever or just aimlessly type words onto your screen? *Zero*Claw, not OpenClaw, is cited in Chris's work. And Chris's work is designed as a research engine. I could argue the same about Andrej's work being a clone of OpenClaw. Chris has receipts, all he wants is this to be properly cited. I think that's pretty graceful of him being the circumstances of plagiarism. https://deepwiki.com/topherchris420/james_library

**Comment by jsegov** [comment]
Nice! I also put together a fork specifically optimized to support RTX 30/40/50 series cards: [jsegov/autoresearch-win-rtx](https://github.com/jsegov/autoresearch-win-rtx).

**Comment by fayerman-source** [comment]
Nice, this is helpful.

I ended up taking a slightly different route on a WSL2 + RTX 5070 12 GB setup:
https://github.com/fayerman-source/autoresearch

I tried to keep the upstream shape mostly intact and just add a few practical things:
- SDPA fallback on non-Hopper GPUs
- small presets (`AR_PRESET=8gb`, `12gb`, `h100`)
- env-var overrides for the main runtime knobs
- a tiny `results.tsv` logger

That was enough to get a full 5-minute run to complete on my side:
- `val_bpb: 1.225821`
- `peak_vram_mb: 3382.3`

Your fork looks more Windows/consumer-RTX focused, which feels complementary. It might be useful if these forks converge a bit on a small shared operator surface so people on normal GPUs don’t end up with totally different setup stories.

**Comment by aniruddhaadak80** [comment]
That's fantastic news. The RTX 5070 seems like it packs a serious compute punch, but fitting everything into 12GB of VRAM must have required a lot of care with the batch sizing and memory management.

Could you share if you had to aggressively lower the sequence length or use heavy gradient checkpointing to fit the optimizer states securely? Really impressive to see it scaled down so efficiently to standard desktop gear.

**Reply by fayerman-source** [reply]
Sequence length did not need to be reduced, and heavy gradient checkpointing was not needed.

What seemed to matter more was making the run shape safer for a consumer GPU:
- fall back to SDPA on non-Hopper GPUs
- use smaller batch defaults / presets
- keep compile off on the safer path

So the fix was mostly about backend/default assumptions, not squeezing memory with checkpointing.

**Comment by fayerman-source** [comment]
Follow-up from the 5070 12 GB path after pushing on the SDPA fallback side a bit more.

I ran a fixed 5-run validation sequence on a fork and got:
- baseline (`SSSL`): `val_bpb 1.269448`, ~79k tok/s
- `WINDOW_PATTERN=L`: `1.189742`, ~105k tok/s
- `+ WARMDOWN_RATIO=0.75`: `1.187303`, ~105k tok/s
- `+ FINAL_LR_FRAC=0.05`: `1.183502`, ~106k tok/s
- `+ small AdamW WD on embeddings/value embeds/lm_head`: `1.184761`, ~103k tok/s

So the big win on this card was also `WINDOW_PATTERN=L` on SDPA, and warmdown `0.75` plus a small LR floor improved a bit more. The targeted extra AdamW decay mix did not reproduce as a win here.

I also opened PR #161 upstream for the code side: non-Hopper SDPA fallback plus cached sliding-window masks. https://github.com/karpathy/autoresearch/pull/161
